### PR TITLE
PUBDEV-6188 - Amazon S3 fetching with credentials in URL

### PR DESCRIPTION
**Pull request by Pscheidl**
[PUBDEV-6188](https://0xdata.atlassian.net/browse/PUBDEV-6188)

This is the **S3** PR, not S3A. This PR touches **PersistS3**, not PersistHDFS.

### CODE REVIEW

**Comment by Pscheidl** [comment]
TODO: Client caching strategy. As credentials are supplied via URI, a new client is built with static credential provider. This means the bucket cache must also contain caches. We should also make sure those are destroyed over time. Could be a significant resource consumer for long-running instances with data loaded from lots of different buckets (client re-init time after time is not really bad, main point of the cache is not to call S3 while user is selecting the file e.g. in Flow).

**Comment by Pscheidl** [comment]
Sidenote: Unlike Hadoop's wrapper, Amazon's client is able to handle special characters just fine, no need to do encoding/decoding on our own.

File: h2o-persist-s3/src/main/java/water/persist/PersistS3.java [hunk 1]
```diff
@@ -25,35 +28,74 @@
...
+    if (accessKeyId != null && accessSecretKey != null) {
+      StringBuilder keyDigestBuilder = new StringBuilder(accessKeyId);
+      keyDigestBuilder.append(accessSecretKey);
+      digest = _messageDigest.digest(keyDigestBuilder.toString().getBytes());
+    } else {
+      digest = null;
+    }
+
+    synchronized (_lock) {
+      if (_s3 == null || digest != _credentialsDigest) {
```
Comment (michalkurka) [inline h2o-persist-s3/src/main/java/water/persist/PersistS3.java]: `_credentialsDigest ` is never updated and stays `null` in this PR.

File: h2o-persist-s3/src/test/java/water/persist/PersistS3Test.java [hunk 2]
```diff
@@ -9,17 +9,21 @@
...
  * Created by tomas on 6/27/16.
  */
 public class PersistS3Test extends TestUtil {
-
+  private static final String AWS_ACCESS_KEY_PROPERTY_NAME = "AWS_ACCESS_KEY_ID";
+  private static final String AWS_SECRET_KEY_PROPERTY_NAME = "AWS_SECRET_ACCESS_KEY";
 
   @BeforeClass
-  public static void setup() { stall_till_cloudsize(5); }
+  public static void setup() { stall_till_cloudsize(1); }
```
Comment (michalkurka) [inline h2o-persist-s3/src/test/java/water/persist/PersistS3Test.java]: I would keep 5 there

File: h2o-persist-s3/src/test/java/water/persist/PersistS3Test.java [hunk 3]
```diff
@@ -9,17 +9,21 @@
...
 import static junit.framework.TestCase.assertEquals;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assume.assumeTrue;
 
 /**
  * Created by tomas on 6/27/16.
  */
 public class PersistS3Test extends TestUtil {
-
+  private static final String AWS_ACCESS_KEY_PROPERTY_NAME = "AWS_ACCESS_KEY_ID";
```
Comment (michalkurka) [inline h2o-persist-s3/src/test/java/water/persist/PersistS3Test.java]: The names of these properties coincide with the names AWS libraries are using to lookup the credentials.

That is why test `testS3UriToKeyUriCredentials` doesn't actually prove that credentials encoded in URL will work (the S3 libraries might take it from the system properties). Aka if I intentionally remove the credentials the test will still pass.

File: h2o-persist-s3/src/test/java/water/persist/PersistS3Test.java [hunk 4]
```diff
@@ -57,4 +61,89 @@ public void testS3Import() throws Exception {
       Scope.exit();
     }
   }
+
+  @Test
+  public void testS3UriToKeyUriCredentials() throws Exception {
+    // This test is only runnable in environment with Amazon credentials properly set {AWS_ACCESS_KEY_ID, AWS_SECRET_ACCESS_KEY}
+    final String accessKey = System.getenv(AWS_ACCESS_KEY_PROPERTY_NAME);
```
Comment (michalkurka) [inline h2o-persist-s3/src/test/java/water/persist/PersistS3Test.java]: Not a good idea imho, reasoning please see above ^^^

Comment (michalkurka) [reply h2o-persist-s3/src/test/java/water/persist/PersistS3Test.java]: Ideally, the test should also show that the import won't work without the credentials in the URL. When a test depends on an external system (or generally has an external assumption) it is good to validate the assumptions before running the test.

File: h2o-persist-s3/src/main/java/water/persist/PersistS3.java (hunk 1 above)
Comment (Pscheidl) [reply h2o-persist-s3/src/main/java/water/persist/PersistS3.java]: I know. Latest patch not applied. WIP.

**Pull request closed without merging**